src/libmongoc/tests/test-mongoc-loadbalanced.c

```diff
@@ -549,6 +550,251 @@ test_loadbalanced_handshake_rejects_non_loadbalanced (void)
    mock_server_destroy (server);
 }
 
+/* Test that an error before the MongoDB handshake completes does NOT go through
+ * SDAM error handling flow. */
+static void
+test_pre_handshake_error_does_not_clear_pool (void)
+{
+   mock_server_t *server;
+   mongoc_uri_t *uri;
+   mongoc_client_pool_t *pool;
+   mongoc_client_t *client_1;
+   mongoc_client_t *client_2;
+   future_t *future;
+   request_t *request;
+   bson_error_t error;
+
+   server = mock_server_new ();
+   mock_server_auto_endsessions (server);
+   mock_server_run (server);
+   uri = mongoc_uri_copy (mock_server_get_uri (server));
+   mongoc_uri_set_option_as_bool (uri, MONGOC_URI_LOADBALANCED, true);
+   pool = mongoc_client_pool_new (uri);
+   client_1 = mongoc_client_pool_pop (pool);
+   client_2 = mongoc_client_pool_pop (pool);
+
+   /* client_1 opens a new connection to send "ping" */
+   future = future_client_command_simple (client_1,
+                                          "admin",
+                                          tmp_bson ("{'ping': 1}"),
+                                          NULL /* read prefs */,
+                                          NULL /* reply */,
+                                          &error);
+   /* A new connection is opened. */
+   request =
+      mock_server_receives_legacy_hello (server, "{'loadBalanced': true}");
+   BSON_ASSERT (request);
+   mock_server_replies_simple (request, LB_HELLO);
+   request_destroy (request);
+   /* The "ping" command is sent. */
+   request = mock_server_receives_msg (server, 0, tmp_bson ("{'ping': 1}"));
+   BSON_ASSERT (request);
+   mock_server_replies_ok_and_destroys (request);
+   ASSERT_OR_PRINT (future_get_bool (future), error);
+   future_destroy (future);
+
+   /* client_2 attempts to open a new connection, but receives an error on the
+    * handshake. */
+   future = future_client_command_simple (client_2,
+                                          "admin",
+                                          tmp_bson ("{'ping': 1}"),
+                                          NULL /* read prefs */,
+                                          NULL /* reply */,
+                                          &error);
+   /* A new connection is opened. */
+   request =
+      mock_server_receives_legacy_hello (server, "{'loadBalanced': true}");
+   BSON_ASSERT (request);
+   capture_logs (true); /* hide Failed to buffer logs. */
```

Comment:
Note, https://github.com/mongodb/mongo-c-driver/blob/4919d4f0e5ff7f4f7a18d433de88e5c16e1e3290/.evergreen/compile-unix.sh#L293:L307 checks the tests that are run alongside compilation (mock server and unit tests) to assert that nothing is logged to stderr. When `mock_server_hangs_up` closes the connection, libmongoc logs the failed socket write. This captures and ignores those logs.